internal/querynodev2/tasks/concurrent_safe_scheduler.go

```diff
@@ -0,0 +1,246 @@
+	"fmt"
+
+	"github.com/milvus-io/milvus/internal/querynodev2/collector"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/metrics"
+	"github.com/milvus-io/milvus/pkg/util/conc"
+	"github.com/milvus-io/milvus/pkg/util/metricsinfo"
+	"github.com/milvus-io/milvus/pkg/util/paramtable"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+const (
+	maxReceiveChanBatchConsumeNum = 100
+)
+
+// newScheduler create a scheduler with given schedule policy.
+func newScheduler(policy schedulePolicy) Scheduler {
+	maxReadConcurrency := paramtable.Get().QueryNodeCfg.MaxReadConcurrency.GetAsInt()
+	maxReceiveChanSize := paramtable.Get().QueryNodeCfg.MaxReceiveChanSize.GetAsInt()
+	log.Info("query node use concurrent safe scheduler", zap.Int("max_concurrency", maxReadConcurrency))
+	return &scheduler{
+		policy:           policy,
+		receiveChan:      make(chan addTaskReq, maxReceiveChanSize),
+		execChan:         make(chan Task),
+		pool:             conc.NewPool[any](maxReadConcurrency, conc.WithPreAlloc(true)),
+		schedulerCounter: schedulerCounter{},
+	}
+}
+
+type addTaskReq struct {
+	task Task
+	err  chan<- error
+}
+
+// scheduler is a general concurrent safe scheduler implementation by wrapping a schedule policy.
+type scheduler struct {
+	policy      schedulePolicy
+	receiveChan chan addTaskReq
+	execChan    chan Task
+	pool        *conc.Pool[any]
+	schedulerCounter
+}
+
+// Add a new task into scheduler,
+// error will be returned if scheduler reaches some limit.
+func (s *scheduler) Add(task Task) (err error) {
+	errCh := make(chan error, 1)
+
+	// TODO: add operation should be fast, is UnsolveLen metric unnesscery?
+	metrics.QueryNodeReadTaskUnsolveLen.WithLabelValues(fmt.Sprint(paramtable.GetNodeID())).Inc()
+
+	// start a new in queue span and send task to add chan
+	s.receiveChan <- addTaskReq{
+		task: task,
+		err:  errCh,
+	}
+	err = <-errCh
+
+	metrics.QueryNodeReadTaskUnsolveLen.WithLabelValues(fmt.Sprint(paramtable.GetNodeID())).Dec()
+	return
+}
+
+// Start schedule the owned task asynchronously and continuously.
+// Start should be only call once.
+func (s *scheduler) Start(ctx context.Context) {
+	// Start a background task executing loop.
+	go s.exec(ctx)
+
+	// Begin to schedule tasks.
+	go s.schedule(ctx)
+}
+
+// schedule the owned task asynchronously and continuously.
+func (s *scheduler) schedule(ctx context.Context) {
+	var task Task
+	for {
+		s.setupReadyLenMetric()
+
+		var execChan chan Task
+		task, execChan = s.setupExecListener(task)
+
+		select {
+		case <-ctx.Done():
+			log.Warn("unexpected quit of schedule loop")
+			return
+		case req := <-s.receiveChan:
+			// Receive add operation request and return the process result.
+			// And consume recv chan as much as possible.
+			s.consumeRecvChan(req, maxReceiveChanBatchConsumeNum)
+		case execChan <- task:
+			// Task sent, drop the ownership of sent task.
+			// And produce new task into execChan as much as possible.
+			task = s.produceExecChan()
```

Comment:
If a task is return, it may be overwritten by line 85 and cause task go orphan

Reply:
setupExecListener (function at line 85) only overwrite task if the task == nil.
So task waiting to send execChan will never go orphan.